### ISSUE TITLE
Copy Paste Rows in Table Editor

### ISSUE THREAD

**Issue by randallnewman** [microsoft/azuredatastudio#14034]
Issue Type: <b>Feature Request</b>

Can't copy and paste a row in editor view. Is there actually no way to copy and paste a row from the table editor?  This is one of the most basic features in MSSMS.  I hope I am missing something.  

Azure Data Studio version: azuredatastudio 1.25.2 (4d4917d328ab7a7526817a0f0ef03237a79aa54d, 2021-01-20T22:52:13.492Z)
OS version: Windows_NT x64 10.0.14393

**Comment by alanrenmsft** [comment]
@randallnewman could you please add some descriptions of what you are trying to achieve? maybe some screenshots?

**Comment by randallnewman** [comment]
Is it possible to put up screen videos. Will be most effective for me to
explain that way.

**Comment by alanrenmsft** [comment]
@randallnewman yes, but you have to open the issue on github in order to attach files.

**Comment by alanrenmsft** [comment]
![image](https://user-images.githubusercontent.com/13777222/105788412-53df9880-5f35-11eb-9302-2dce7f80c66a.png)

**Comment by randallnewman** [comment]
See video:

https://user-images.githubusercontent.com/4906943/105862076-66051980-5fbd-11eb-9f33-861cc2af3837.mp4

**Comment by randallnewman** [comment]
See video above where I show how I can copy paste in MSSMS but not in Azure

**Comment by alanrenmsft** [comment]
@randallnewman thanks for the video, now it is very clear what you are asking for :) this is the edit data feature. let me assign this to the correct area and let the area owner take a look.

**Comment by randallnewman** [comment]
Great. If you find this helpful, I can add more items as I find them.

**Comment by plmosqueda** [comment]
Any advance on this??

Thank you all

**Comment by kishanhickman** [comment]
Is there an update on this?  its the only reason I have SSMS still installed on my laptop!  

Also, when you copy an item from a cell, the formatting is not maintained (which it is on SSMS) - again another annoying problem that keeps me still having to use SSMS. 

For exmaple, say that I have code stored in the DB and I want to copy that to Visual Studio Code, the formatting is lost and its just one long line of code.  In SSMS the formatting is maintained so I can copy into VS Code, change something and copy and paste it back.  With Data Studio this isnt possible and really frustrating.

**Comment by randallnewman** [comment]
I have not heard an update.  I am still using SSMS too.  I feel like I'm still in Windows 98.  Come on MS!

**Comment by dv10den** [comment]
Just realized this issue where opened 2 years ago... I really want this feature

**Comment by joergrosenhagen** [comment]
Any Updates on this?

**Comment by alanrenmsft** [comment]
@erinstellato-ms @kburtram  FYI.

**Comment by erinstellato-ms** [comment]
@kishanhickman to confirm - if you had this functionality in ADS, then you would be able to do 100% of the work you need in ADS, and not use SSMS again?  This is the ONLY thing blocking you?  Perfectly fine if so, I just want to make sure I understand priority of functionality based on your use and workflow.  Also, this is when you are editing data in a table, correct?  Not from the output pane after running query (just want to verify, thanks).

@randallnewman I want to make sure I understand your use case, and I read through the comments so forgive me if I'm repeating information you've already supplied.  You are editing data in a table, and then you also want to copy values from the editable cells, and paste them elsewhere, correct?  If that is correct, what is the workflow?  I'm looking to understand why you copy the data from the edit data window versus from the output window (e.g. run a query and copy from there).

For anyone else (who I did not tag) who has commented or upvoted, please feel free to add another comment with your use scenario/workflow.  I understand you want to copy the data, I would like to also understand how you get to this window and why.  What are the steps you go through and what problem do you solve with this method?  Thank you!

**Comment by joergrosenhagen** [comment]
Hello @erinstellato-ms ,
in our case it is about simply copying a line in edit mode and pasting it directly again. So, it can be duplicated (if necessary, without the self-counting Id, which is assigned automatically) and then edited again before the data is sent to the SQL.

The use case is just for people who can't write SQL statements and then only "just" want to change something directly in the table.

**Comment by kishanhickman** [comment]
Hi Erin,

Yes that is main sticking point that prevents from using ADS

Best Regards,


*Kishan Hickman*

**Comment by erinstellato-ms** [comment]
> Hello @erinstellato-ms , in our case it is about simply copying a line in edit mode and pasting it directly again. So, it can be duplicated (if necessary, without the self-counting Id, which is assigned automatically) and then edited again before the data is sent to the SQL.
> 
> The use case is just for people who can't write SQL statements and then only "just" want to change something directly in the table.

Thank you for providing that additional detail @joergrosenhagen, that's very helpful.  And for those users who do not write T-SQL statements, is ADS a more preferable tool overall for them?  My assumption (possibly incorrect) is that they are not DBAs, even though they have access to modify data.  Correct?  I would like to understand more about what those users need to do in a tool like ADS or SSMS.  I really appreciate your reply and help here!

**Comment by joergrosenhagen** [comment]
> And for those users who do not write T-SQL statements, is ADS a more preferable tool overall for them? 

ADS is faster and more lightweight, so yes.

> My assumption (possibly incorrect) is that they are not DBAs, even though they have access to modify data. Correct? 

This is correct. I am also aware that this should not be the way how someone changes or adds data, but it just works in SSMS and is so wished.

> I would like to understand more about what those users need to do in a tool like ADS or SSMS. I really appreciate your reply and help here!

I am happy that I can help

**Comment by kkarasti** [comment]
I'm a software developer and I create database tables for the development phase of an app. I use C# for the app and am using LINQ so I don't write T-SQL statements.

Copy/Paste would be wonderful so I can use ADS exclusively.

The only other thing that I sometimes do is create views. The graphical view creation in SSMS is nice and I don't see that in ADS. But I could Script As Create an existing view in ADS and modify and execute the SQL to create a view that way. This functionality is probably a bigger task than the copy/paste and I don't do this often.

**Comment by erinstellato-ms** [comment]
> 

Hi @kkarasti - thank you for sharing your use case and the experience of using LINQ and not writing many T-SQL statements.  When you use Copy/Paste, is it because you are populating just a bit of data into the table (e.g. a few rows, not thousands of rows)?  I want to understand the number of rows (volume of data) you might be copying and pasting in this scenario.  Thus far, folks have mentioned one row at a time.  Is that your scenario, or do you need/want to copy more rows (and how many)?

I understand creating views, and you are correct in that the view designer does not exist in ADS.  We haven't seen quite the same demand for that as table designer, so it is lower on our list.  Your option to script out an existing view and then modify it is a good workaround for the time being.   To upvote the existing issue for a view designer, please see:  https://github.com/microsoft/azuredatastudio/issues/7310

Thanks!

**Comment by kkarasti** [comment]
Correct, just a bit of data. Usually the tables that I'm using are storing configuration data that I only read from. So I may want to add a duplicate record to see how my app handles that situation. Or a duplicate record and then modify one small part of it.

**Comment by bassem-mf** [comment]
This feature will help me copy/paste data between tables that have the same columns. It will also help me copy data from a spreadsheet and paste to insert into a table.

**Comment by imamulnayeem** [comment]
ADS has so many amazing features how come it missed this one!! I'm really looking forward to have this feature on ADS.

**Comment by cheenamalhotra** [comment]
### Proposed Solution

1. Support new context menu option on existing rows to Copy or use Keyboard Shortcut (Ctrl + C):
    <img src="https://github.com/microsoft/azuredatastudio/assets/13396919/1d18c2a5-3ede-413b-943a-74503a824b29" width=450 />

2. Support new context menu option on empty row to Paste or use Keyboard Shortcut (Ctrl + V)

cc @erinstellato-ms

**Comment by imamulnayeem** [comment]
> ### Proposed Solution
> 1. Support new context menu option on existing rows to Copy or use Keyboard Shortcut (Ctrl + C):
>    <img alt="" width="450" src="https://user-images.githubusercontent.com/13396919/270812029-1d18c2a5-3ede-413b-943a-74503a824b29.png">
> 2. Support new context menu option on empty row to Paste or use Keyboard Shortcut (Ctrl + V)
> 
> cc @erinstellato-ms

I completely agree with you.

**Comment by GearTheWorld** [comment]
2024-1-18 : Unable to Copy / Paste rows in the editor yet ?  We can Copy a row from a Result Set but not paste it in the editor right now.  This is the most useful thing in SSMS

Version: 1.47.1 (user setup)
Commit: b6f7beb01f92adaa4b79b6b6f3ac704e95cafe6e
Date: 2024-01-10T02:15:02.628Z
VS Code: 1.82.0
Electron: 25.8.4
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Windows_NT x64 10.0.22621

**Comment by Prologh** [comment]
Bumping this issue. Even Visual Studio SQL Server Object Explorer supports pasting whole tables.

**Comment by jjamid** [comment]
Azure Data Studio is amazing, but without these basic features it's quite useless... We still have an work with a SSMS and no point switching back and forth.

The fact that the first recation was "show me in a video what you are trying to do" for something so basic and obvious says it all.
Three years and this absolute MUST-HAVE feature is no where to be seen. 
Seems that the people who are in charge of this product are not people who do DB work themselves and don't see the inconvenieces. 

This happens with the Snipping Tool too and many other tools, years and years and still no "Add text DIRECTLY on the screenshot" (not draw.. add clear text). Instead, you still have to use a second tool to add that text.

**Comment by teeinep** [comment]
SSMS will most times freeze up when trying to edit an Azure table, so I thought I'd give ADS a try.  Nope.  I can copy a row but cannot paste the row back into the table.

My use case is - I have a table where I track the history of an application and when I create a new version (Major, Minor, Build, Revision) I need to add a new row and document the bug fix in the comment.

I don't have time to write a new UI to do this so I'm going back to SSMS where I can, sometimes, copy and paste a row in a table.  It's a crap shoot whether it freezes up or not.

**Comment by erinstellato-ms** [comment]
> SSMS will most times freeze up when trying to edit an Azure table, so I thought I'd give ADS a try. Nope. I can copy a row but cannot paste the row back into the table.
> 
> My use case is - I have a table where I track the history of an application and when I create a new version (Major, Minor, Build, Revision) I need to add a new row and document the bug fix in the comment.
> 
> I don't have time to write a new UI to do this so I'm going back to SSMS where I can, sometimes, copy and paste a row in a table. It's a crap shoot whether it freezes up or not.

@teeinep You copy the entire row in SSMS (using CTRL + C) and then paste the entire row (using CTRL + V)?  Are you doing Edit Top 200 rows on the table, and then copying and pasting from there?

**Comment by teeinep** [comment]
Yes, Edit Top 200, CTRL+C, CTRL+V.  
But in most cases, SSMS will freeze up when choosing Edit Top 200 from the table context menu.  A popup message appears "SSMS is busy..." or something to that effect, along with the hourglass.  SSMS never recovers, and Task Manager is required to terminate the process.

In other cases the freeze up happens when pasting the copied row into the >* row (new row) at the bottom of the table editor.

The is directly related to the number of query windows open in SSMS because with a fresh instance of SSMS I'm usually lucky and get my edits done.  But if I'm in the middle of something and I have a lot of windows open (and a lot of work to lose), I'll most likely get locked up when selecting Edit Top 200.  So maybe there's a memory issue there.

Tony Liguori
***@***.*** 
 

    On Wednesday, August 14, 2024 at 11:45:01 AM CDT, Erin Stellato ***@***.***> wrote:  
 
 



SSMS will most times freeze up when trying to edit an Azure table, so I thought I'd give ADS a try. Nope. I can copy a row but cannot paste the row back into the table.

My use case is - I have a table where I track the history of an application and when I create a new version (Major, Minor, Build, Revision) I need to add a new row and document the bug fix in the comment.

I don't have time to write a new UI to do this so I'm going back to SSMS where I can, sometimes, copy and paste a row in a table. It's a crap shoot whether it freezes up or not.


@teeinep You copy the entire row in SSMS (using CTRL + C) and then paste the entire row (using CTRL + V)? Are you doing Edit Top 200 rows on the table, and then copying and pasting from there?

**Comment by RamblingGeekUK** [comment]
As others have stated, this is annoying.  I'm using ADS for black theme, ability to increase UI size on 4k screen... however not being able to past rows in is a pain. Seems like a minimum requirement.

**Comment by HamzehJankhout-W** [comment]
I don't understand why adding features that are great in Azure Studio that should have been in SSMS and vice versa,
Adding multiple rows is a basic feature and it has been 3 years and still nothing.

**Comment by JDBC-0** [comment]
+1   .... that's also the number 1 reason we still have to work with SSMS which freezes about 3 times a working day ...

**Comment by KAK-ABT** [comment]
Well now that ADS is being retired, this feature will never be, and it isn't a feature in VSCode yet. Here is my solution for duplicating records using LINQPad.

Create a C# program as follows:

```
void Main()
{
	var original = (from p in KAKs where p.ID == 15 select p);
	var source = original.First();
	var cloned = Clone(source);
	cloned.ID = 0;
	//cloned.TestResult = "TEST1";
	KAKs.InsertOnSubmit(cloned);
	SubmitChanges();
}

public static T Clone<T>(T source)
{
	var dcs = new System.Runtime.Serialization.DataContractSerializer(typeof(T));
	using (var ms = new System.IO.MemoryStream())
	{
		dcs.WriteObject(ms, source);
		ms.Seek(0, System.IO.SeekOrigin.Begin);
		return (T)dcs.ReadObject(ms);
	}
}
```
The commented line above is optional if you want to change some data in the new record.
